ThorsAngerVaNeT/nodejs-aws-shop-be
# Task 9 - Backend For Frontend

1. Task: https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/9_backend_for_frontend/task.md
2. 
    - Product Service API URL - https://v7epzfc1vd.execute-api.eu-west-1.amazonaws.com/dev/products
    - Cart Service API URL - https://hwh9egqhs6.execute-api.eu-west-1.amazonaws.com/api/profile/cart
    - BFF Service API URL - http://thorsangervanet-bff-api-dev.eu-west-1.elasticbeanstalk.com/
      - [/products](http://thorsangervanet-bff-api-dev.eu-west-1.elasticbeanstalk.com/products) - equivalent of Product Service API URL provided above
      - [/cart](http://thorsangervanet-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart) - equivalent of Cart Service API URL 
  ~~- NestJS BFF Service API URL - http://thorsangervanet-bff-nest-api-dev.eu-west-1.elasticbeanstalk.com/~~ Terminated it to prevent getting out of free tier limits
    ~~- [/products](http://thorsangervanet-bff-nest-api-dev.eu-west-1.elasticbeanstalk.com/products) - equivalent of Product Service API URL provided above~~
    ~~- [/cart](http://thorsangervanet-bff-nest-api-dev.eu-west-1.elasticbeanstalk.com/cart) - equivalent of Cart Service API URL provided above~~
    - BE Repo https://github.com/ThorsAngerVaNeT/nodejs-aws-shop-be
3. FE deploy https://dkct2oxjblzy1.cloudfront.net/
FE PR https://github.com/ThorsAngerVaNeT/nodejs-aws-shop-fe/pull/9
FE Repo https://github.com/ThorsAngerVaNeT/nodejs-aws-shop-fe
6. Done 14.08.2022 / deadline 14.08.2022
7. Score: 7/7
    #### Evaluation criteria (each mark includes previous mark criteria)
    - [x] A working and correct Express application should be in the bff-service folder. Reviewer can start this application locally with any valid configuration in the .env file and this application should works as described in the Task 9.1
    
      You can test createProduct lambda with following URL http://thorsangervanet-bff-api-dev.eu-west-1.elasticbeanstalk.com/products, send POST request with that JSON body:
        ```json
        {
          "count": 4,
          "description": "Testing Cache Desc",
          "last": "Testing Cache Last",
          "sole": "Testing Cache Sole",
          "price": 196,
          "title": "Testing Cache Title",
          "image_url": "https://d1wh94yq5jmkrj.cloudfront.net/ce80facb-9804-489b-9b9a-b3a1045df2bd/01.jpg"
        }
        ```
        To check that error response also the same as at Product Service you can delete any required field from JSON (title,, last, sole, count) and response will look like this:
        ```json
          {
              "error": "null value in column \"count\" of relation \"stocks\" violates not-null constraint"
          }
        ```
        or send GET request to Product Service with not empty body :)
    
    #### Additional (optional) tasks
    - [x] Add a cache at the BFF Service level for a request to the `getProductsList` lambda function of the Product Service. The cache should expire in 2 minutes. 
      **Please keep in mind that I added check for empty body to prevent caching error**

      How to test:
      - Get products list
      - Create new product
      - Get products list - result shouldn’t have new product
      - Wait more than 2 minutes
      - Get products list - result should have new product

# Plesase don't hesitate to contact me in Discord if you have any questions. - VaNeT#5916
